File: hall_of_fame/wi-seong-cheol/BOJ/0x07 Deque/10866.cpp
//
//  10866.cpp
//  main
//
//  Created by wi_seong on 2022/11/10.
//

/*
 [Input]
 15
 push_back 1
 push_front 2
 front
 back
 size
 empty
 pop_front
 pop_back
 pop_front
 size
 empty
 pop_back
 push_front 3
 empty
 front
 [Output]
 2
 1
 2
 0
 2
 1
 -1
 0
 1
 -1
 0
 3
 */
// 시간복잡도: O(n)
// 최악시간: 10,000
// 난이도: Silver 4
// Timer: 5m
// Url: https://www.acmicpc.net/problem/10886

#include <iostream>
#include <queue>
#include <string>

using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    int t; cin >>t;
    deque<int> dq;
    
    while(t--) {
        string op; cin >> op;
        if(op == "push_front") {
            int x; cin >> x;
            dq.push_front(x);
        } else if(op == "push_back") {
            int x; cin >> x;
            dq.push_back(x);
        } else if(op == "pop_front") {
            if(dq.empty()) cout << "-1\n";
            else {
                cout << dq.front() << '\n';
                dq.pop_front();
            }
        } else if(op == "pop_back") {
            if(dq.empty()) cout << "-1\n";
            else {
                cout << dq.back() << '\n';
                dq.pop_back();
            }
        } else if(op == "size") {
            cout << dq.size() << '\n';
        } else if(op == "empty") {
            cout << dq.empty() << '\n';
        } else if(op == "front") {
            if(dq.empty()) cout << "-1\n";
            else cout << dq.front() << '\n';
        } else {
            if(dq.empty()) cout << "-1\n";
            else cout << dq.back() << '\n';
        }
    }
    
    return 0;
}

File: ioloolo/BOJ/SolvedAc Class/Class 2/10866.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.Deque;
import java.util.LinkedList;
import java.util.StringTokenizer;

public class Main {
	BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
	BufferedWriter writer = new BufferedWriter(new OutputStreamWriter(System.out));

	StringTokenizer tokenizer;

	int N;
	Operator[] arr;

	void input() throws IOException {
		N = Integer.parseInt(reader.readLine());

		arr = new Operator[N];

		for (int i = 0; i < N; ++i) {
			tokenizer = new StringTokenizer(reader.readLine());

			arr[i] = new Operator();
			arr[i].type = Operator.Type.valueOf(tokenizer.nextToken().toUpperCase());

			if (tokenizer.hasMoreTokens()) {
				arr[i].operand = Integer.parseInt(tokenizer.nextToken());
			}
		}
	}

	void solve() throws IOException {
		Deque<Integer> queue = new LinkedList<>();

		for (Operator operator : arr) {
			if (operator.type == Operator.Type.PUSH_FRONT) {
				queue.addFirst(operator.operand);
				continue;
			} else if (operator.type == Operator.Type.PUSH_BACK) {
				queue.addLast(operator.operand);
				continue;
			} else if (operator.type == Operator.Type.POP_FRONT) {
				if (queue.isEmpty()) {
					writer.write("-1");
				} else {
					writer.write(Integer.toString(queue.pollFirst()));
				}
			} else if (operator.type == Operator.Type.POP_BACK) {
				if (queue.isEmpty()) {
					writer.write("-1");
				} else {
					writer.write(Integer.toString(queue.pollLast()));
				}
			} else if (operator.type == Operator.Type.SIZE) {
				writer.write(Integer.toString(queue.size()));
			} else if (operator.type == Operator.Type.EMPTY) {
				writer.write(Integer.toString(queue.isEmpty() ? 1 : 0));
			} else if (operator.type == Operator.Type.FRONT) {
				if (queue.isEmpty()) {
					writer.write("-1");
				} else {
					writer.write(Integer.toString(queue.getFirst()));
				}
			} else if (operator.type == Operator.Type.BACK) {
				if (queue.isEmpty()) {
					writer.write("-1");
				} else {
					writer.write(Integer.toString(queue.getLast()));
				}
			}

			writer.newLine();
		}
	}

	static class Operator {
		public enum Type {
			PUSH_FRONT, PUSH_BACK, POP_FRONT, POP_BACK,
			SIZE, EMPTY, FRONT, BACK;
		}

		Type type;
		int operand;
	}

	public static void main(String[] args) throws IOException {
		Main main = new Main();

		main.input();
		main.solve();

		main.reader.close();
		main.writer.close();
	}
}
